### PR TITLE
:bug: Don't reconcile security groups on NLBs from older versions of CAPA

### LINKED ISSUE #4790: Network Load Balancers won't reconcile in main
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Upgrading from a version of CAPA prior to adding #4733 causes network load balancers to fail reconciliation.

```
E0212 15:21:48.220015       1 awscluster_controller.go:280] "failed to reconcile load balancer" err=<
        failed to apply security groups to load balancer "aws-us-cw-mf-test-1": InvalidConfigurationRequest: You cannot set security groups on a Network Load Balancer which was created without any security groups.
                status code: 400, request id: 05010ea1-3b8f-4c6c-9674-2f785030006e
 > controller="awscluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSCluster" AWSCluster="system-clusters/aws-us-cw-mf-test-1-66z8p" namespace="system-clusters" name="aws-us-cw-mf-test-1-66z8p" reconcileID="8e92c000-83ee-411e-a9d1-6aa72495b6c8" cluster="system-clusters/aws-us-cw-mf-test-1"
```

**What did you expect to happen:**

Network load balancers previously created shouldn't fail to reconcile.

**Anything else you would like to add:**

The network load balancer [documentation](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-security-groups.html#security-group-considerations) states that an NLB created without security groups cannot have security groups added or removed later.

**Environment:**

- Cluster-api-provider-aws version: `master` including #4733 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by nrb**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

The logic introduced in #4733 removed the checks on attaching security groups to network load balancers. This is fine for newly created NLBs, but for ones created in older versions of CAPI, security groups _cannot_ be attached, and we should respect those settings still.

**Which issue(s) this PR fixes**
Fixes #4790 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests

**Release note**:
```release-note
NONE
```

/priority critical-urgent

### CODE REVIEW

File: pkg/cloud/services/elb/loadbalancer_test.go [hunk 1]
```diff
@@ -1847,6 +1847,70 @@ func TestReconcileV2LB(t *testing.T) {
 				}
 			},
 		},
+		{
+			name: "ensure NLB without SGs doesn't attempt to add new SGs",
```
Comment (nrb) [inline pkg/cloud/services/elb/loadbalancer_test.go:1851]: This test is invalid, as it's passing even when the conditions aren't satisfied. It needs to have some more mocks fleshed out in order to be accurate.

/hold

Comment (nrb) [reply pkg/cloud/services/elb/loadbalancer_test.go:1851]: /unhold

File: pkg/cloud/services/elb/loadbalancer.go [hunk 2]
```diff
@@ -1567,3 +1567,15 @@ func chunkELBs(names []string) [][]string {
 	}
 	return chunked
 }
+
+func shouldReconcileSGs(lbType infrav1.LoadBalancerType, existingSGs []string, specSGs []string) bool {
+	// Backwards compat: NetworkLoadBalancers were not always capable of having security groups attached.
+	// Once created without a security group, the NLB can never have any added.
```
Comment (AndiDog) [inline pkg/cloud/services/elb/loadbalancer.go:1573]:
```suggestion
	// Once created without a security group, the NLB can never have any added
	// (https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-security-groups.html).
```

File: pkg/cloud/services/elb/loadbalancer.go [hunk 3]
```diff
@@ -1567,3 +1567,15 @@ func chunkELBs(names []string) [][]string {
...
+	if lbType == infrav1.LoadBalancerTypeNLB && len(existingSGs) == 0 {
+		return false
```
Comment (AndiDog) [inline pkg/cloud/services/elb/loadbalancer.go:1575]: Users won't get a warning in logs if their desired manifest specifies SGs but their existing AWS NLB has none. Can we maybe add a log line for this?

Comment (nrb) [reply pkg/cloud/services/elb/loadbalancer.go:1575]: Done!

**Review by damdo — COMMENTED** [review]
/lgtm
Thanks for the fix.

/assign @richardcase 
For final approval

**Comment by damdo** [comment]
Pod got killed, retrying:

/test pull-cluster-api-provider-aws-e2e

**Comment by damdo** [comment]
Pod got killed, retrying:

/test pull-cluster-api-provider-aws-e2e

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**